### PR TITLE
Yarn 4 support

### PR DESCRIPTION
**Pull request by JLHwung**
This is a breaking change.

We bumped the yarn dependencies and adapted to Yarn 4 breaking changes. Because the plugin relies on yarn internal packages, we will not support Yarn 3 in the near future. Use 0.1.0 if you are using Yarn 3.

The built plugin seems to work with the Babel repo after it upgraded to Yarn 4.

### CODE REVIEW

File: .yarnrc.yml [hunk 1]
```diff
@@ -1,5 +1,7 @@
-plugins:
-  - path: .yarn/plugins/@yarnpkg/plugin-interactive-tools.cjs
-    spec: "@yarnpkg/plugin-interactive-tools"
```
Comment (JLHwung) [inline .yarnrc.yml:3]: The plugin is removed because Yarn 4 includes this plugin.

File: .yarnrc.yml [hunk 2]
```diff
@@ -1,5 +1,7 @@
...
+compressionLevel: mixed
 
-yarnPath: .yarn/releases/yarn-3.6.1.cjs
+enableGlobalCache: false
+
+yarnPath: .yarn/releases/yarn-4.0.0.cjs
+
+nodeLinker: pnp
```
Comment (JLHwung) [inline .yarnrc.yml:7]: It seems to me the default nodeLinker in Yarn 4 is now `node-modules`.

File: .pnp.loader.mjs [hunk 3]
```diff
@@ -1,2047 +0,0 @@
-import { URL as URL$1, fileURLToPath, pathToFileURL } from 'url';
```
Comment (nicolo-ribaudo) [inline .pnp.loader.mjs:1]: Do we need to add this file to .gitignore (like .pnp.cjs), or does Yarn not generate it anymore?

**Review by nicolo-ribaudo — APPROVED** [review]

File: .pnp.loader.mjs (hunk 3 above)
Comment (JLHwung) [reply .pnp.loader.mjs:1]: Yarn 4 does not generate this file anymore, only `.pnp.cjs` is generated. Maybe it has been included in Yarn 4 as well.

**Comment by JLHwung** [comment]
I will open a new PR to Babel using the built version of this PR, if the Babel CI is green it should be good to merge.

File: .yarnrc.yml (hunk 2 above)
Comment (merceyz) [reply .yarnrc.yml:7]: It's not, maybe you set it globally in `~/.yarnrc.yml`?

File: .pnp.loader.mjs (hunk 3 above)
Comment (merceyz) [reply .pnp.loader.mjs:1]: Yarn still generates it if it detects that ESM support is required.

File: .yarnrc.yml (hunk 2 above)
Comment (JLHwung) [reply .yarnrc.yml:7]: Indeed, now I wonder why previous Yarn version generates pnp builds.🤔

**Comment by JLHwung** [comment]
Currently the Babel CI is failing because the serialized [`esmExports` info](https://github.com/nicolo-ribaudo/yarn-plugin-conditions/blob/204bb4fd6b254dcc5ac6b675e3aa3e7c176ea812/sources/conditionUtils.ts#L72) is unfortunately recognized as [multiple cache keys](https://github.com/yarnpkg/berry/blob/abcdaa8b77746753c6128734a3ccc54cee822432/packages/yarnpkg-core/sources/Project.ts#L399) in Yarn 4. I will try to move on to a new delimiter, maybe `|` or even a lone surrogate U+F800 if we ever want to fully support string module names.

**Comment by merceyz** [comment]
That RegExp hasn't changed for 4 years so that's odd, I would expect it to be related to https://github.com/yarnpkg/berry/pull/4305.

**Comment by JLHwung** [comment]
image

I am not familiar with the internals, maybe the CI error is due to the PR as you mentioned. In the debugger the key split is the first step where the descriptor range  `key2` `"...(esm:default,Chalk)"` is incorrectly cropped into `entry` `"...(esm:default"`, which is later thrown by the condition plugin. 

On a second thought, maybe that regex should have been `/ *, +/g`? Since the generated lockfile always separate items by `, `.

I will try your approach, it would be great if we don't have to move on a new serialization. Thank you!

**Comment by nicolo-ribaudo** [comment]
Ty! Is this ready now that CI on the Babel PR is passing?

**Comment by JLHwung** [comment]
> Ty! Is this ready now that CI on the Babel PR is passing?

Yes it is ready for review now. 🙏

File: .yarnrc.yml [hunk 4]
```diff
@@ -1,5 +1,7 @@
-plugins:
-  - path: .yarn/plugins/@yarnpkg/plugin-interactive-tools.cjs
-    spec: "@yarnpkg/plugin-interactive-tools"
+compressionLevel: mixed
 
-yarnPath: .yarn/releases/yarn-3.6.1.cjs
+enableGlobalCache: false
```
Comment (liuxingbaoyu) [inline .yarnrc.yml:3]: Just curious, is there any reason for this?

Comment (JLHwung) [reply .yarnrc.yml:3]: It is automatically generated by Yarn 4 when migrating from Yarn 3 config. The default is now `true` in Yarn 4, but it was `false` in Yarn 3.

**Pull request merged**